homeassistant/components/smarttub/controller.py

```diff
@@ -0,0 +1,123 @@
+"""Interface to the SmartTub API."""
+
+import asyncio
+from datetime import timedelta
+import logging
+
+from aiohttp import client_exceptions
+import async_timeout
+from smarttub import APIError, LoginFailed, SmartTub
+from smarttub.api import Account
+
+from homeassistant.config_entries import SOURCE_REAUTH
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, POLLING_TIMEOUT, SCAN_INTERVAL
+from .helpers import get_spa_name
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SmartTubController:
+    """Interface between Home Assistant and the SmartTub API."""
+
+    def __init__(self, hass):
+        """Initialize an interface to SmartTub."""
+        self._hass = hass
+        self._api = SmartTub(async_get_clientsession(hass))
```

Comment:
Set this to `None` and create it in `async_setup_entry`

http://blog.millermedeiros.com/constructors-should-not-cause-side-effects/

Reply:
We need it in order to call `login` which is called from the config flow to validate the credentials. Do you want me to move it to a new method which gets called from `async_setup_entry` and `login`?

Reply:
Whatever works that gets it out of `__init__` is fine